### PR TITLE
tune api for schedulers

### PR DESCRIPTION
**Pull request by qingyun-wu**
- Change flaml.tune API for schedulers
- Rename "prune_attr" to "resource_attr"
- Add tests for schedulers
- Re-run notebooks that use schedulers

### CODE REVIEW

File: flaml/automl.py [hunk 1]
```diff
@@ -1531,7 +1531,7 @@ def points_to_evaluate(self) -> dict:
         return points
 
     @property
-    def prune_attr(self) -> Optional[str]:
+    def resource_attr(self) -> Optional[str]:
         """Attribute for pruning
```
Comment (sonichi) [inline flaml/automl.py:1535]: update docstr

File: flaml/searcher/blendsearch.py [hunk 2]
```diff
@@ -91,17 +91,17 @@ def __init__(
                 points_to_evaluate.
             time_budget_s: int or float | Time budget in seconds.
             num_samples: int | The number of configs to try.
-            prune_attr: A string of the attribute used for pruning.
+            resource_attr: A string of the attribute used for pruning.
```
Comment (sonichi) [inline flaml/searcher/blendsearch.py:94]: update docstr

File: flaml/tune/tune.py [hunk 3]
```diff
@@ -117,11 +119,12 @@ def run(
...
     points_to_evaluate: Optional[List[dict]] = None,
     evaluated_rewards: Optional[List] = None,
-    prune_attr: Optional[str] = None,
+    resource_attr: Optional[str] = None,
     min_resource: Optional[float] = None,
     max_resource: Optional[float] = None,
     reduction_factor: Optional[float] = None,
-    report_intermediate_result: Optional[bool] = False,
+    scheduler: Optional[Union[TrialScheduler, str]] = None,
+    use_flaml_scheduler: Optional[bool] = False,
```
Comment (sonichi) [inline flaml/tune/tune.py:127]: remove

File: flaml/tune/tune.py [hunk 4]
```diff
@@ -205,21 +208,29 @@ def compute_with_config(config):
...
-            prune_attr; only valid if prune_attr is not in space.
+            resource_attr; only valid if resource_attr is not in space.
         max_resource: A float of the maximal resource to use for the
-            prune_attr; only valid if prune_attr is not in space.
+            resource_attr; only valid if resource_attr is not in space.
         reduction_factor: A float of the reduction factor used for incremental
             pruning.
-        report_intermediate_result: A boolean of whether intermediate results
-            are reported. If so, early stopping and pruning can be used.
+        scheduler: A scheduler for executing the experiment. Can be 'flaml','asha'
```
Comment (sonichi) [inline flaml/tune/tune.py:219]:
```suggestion
        scheduler: A scheduler for executing the experiment. Can be 'flaml', 'asha'

```

File: flaml/tune/tune.py [hunk 5]
```diff
@@ -205,21 +208,29 @@ def compute_with_config(config):
...
+            https://arxiv.org/pdf/1911.04706.pdf).
+            When set 'asha', the input for arguments "resource_attr",
+            "min_resource", "max_resource" and "reduction_factor" will be passed
+            to ASHA's "time_attr",  "max_t", "grace_period" and "reduction_factor"
+            respectively. You can also provide a self-defined scheduler instance
+            of the TrialScheduler class. When 'asha' or self-defined scheduler is
+            used, you usually need to report intermediate results in the training
+            function. Please find examples using different types of schedulers
+            and how to set up the corresponding training functions in
+            test/tune/test_scheduler.py.
```
Comment (sonichi) [inline flaml/tune/tune.py:233]: later, change this to a pointer to notebook.

File: test/tune/test_scheduler.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
+"""Require: pip install flaml[test,ray]
+"""
+from logging import raiseExceptions
+from ray.tune.schedulers import TrialScheduler
```
Comment (sonichi) [inline test/tune/test_scheduler.py:4]: this will fail when ray is not installed

File: test/tune/test_scheduler.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    print("No scheduler, test error:", abs(10 / 2 - best_config["z"] / 2))
+
+
+def test_asha_scheduler():
+    best_config = test_scheduler(scheduler="asha")
+    print("Auto ASHA scheduler, test error:", abs(10 / 2 - best_config["z"] / 2))
+
+
+def test_custom_scheduler():
+    from ray.tune.schedulers import HyperBandScheduler
```
Comment (sonichi) [inline test/tune/test_scheduler.py:106]: fail when ray is not installed.

File: test/tune/test_scheduler.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+    my_scheduler = HyperBandScheduler(
+        time_attr="samplesize", max_t=1000, reduction_factor=2
+    )
+    best_config = test_scheduler(scheduler=my_scheduler)
+    print("Custom ASHA scheduler, test error:", abs(10 / 2 - best_config["z"] / 2))
+
+
+def test_custom_scheduler_default_time_attr():
+    from ray.tune.schedulers import ASHAScheduler
```
Comment (sonichi) [inline test/tune/test_scheduler.py:116]: fail when ray is not installed.

File: notebook/flaml_pytorch_cifar10.ipynb [hunk 9]
```diff
@@ -301,9 +337,225 @@
...
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "2021-12-02 20:35:15,350\tERROR tune.py:557 -- Trials did not complete: [train_cifar_34fb4d64]\n",
```
Comment (sonichi) [inline notebook/flaml_pytorch_cifar10.ipynb:553]: There is an error in execution.
Could you clear the output of this notebook? Like the original.

File: flaml/tune/tune.py [hunk 10]
```diff
@@ -12,11 +12,13 @@
 
     assert ray_version >= "1.0.0"
     from ray.tune.analysis import ExperimentAnalysis as EA
+    from ray.tune.schedulers import TrialScheduler
 
     ray_import = True
 except (ImportError, AssertionError):
     ray_import = False
     from .analysis import ExperimentAnalysis as EA
+    from ..scheduler.trial_scheduler import TrialScheduler
```
Comment (sonichi) [inline flaml/tune/tune.py:21]: causing looped import which fails the test.

File: flaml/searcher/blendsearch.py [hunk 11]
```diff
@@ -91,17 +91,9 @@ def __init__(
...
-                When resource_attr is in space, it is a hyperparameter, e.g.,
-                    'n_iters', and the best value is unknown.
-                When resource_attr is not in space, it is a resource dimension,
-                    e.g., 'sample_size', and the peak performance is assumed
-                    to be at the max_resource.
-            min_resource: A float of the minimal resource to use for the
-                resource_attr; only valid if resource_attr is not in space.
-            max_resource: A float of the maximal resource to use for the
-                resource_attr; only valid if resource_attr is not in space.
+            resource_attr: A string to specify the resource dimension.
```
Comment (sonichi) [inline flaml/searcher/blendsearch.py:94]: and the best performance is assumed to be at the max_resource?

File: flaml/searcher/flow2.py [hunk 12]
```diff
@@ -67,17 +67,9 @@ def __init__(
...
-                When resource_attr is in space, it is a hyperparameter, e.g.,
-                    'n_iters', and the best value is unknown.
-                When resource_attr is not in space, it is a resource dimension,
-                    e.g., 'sample_size', and the peak performance is assumed
-                    to be at the max_resource.
-            min_resource: A float of the minimal resource to use for the
-                resource_attr; only valid if resource_attr is not in space.
-            max_resource: A float of the maximal resource to use for the
-                resource_attr; only valid if resource_attr is not in space.
+            resource_attr: A string to specify the resource dimension.
```
Comment (sonichi) [inline flaml/searcher/flow2.py:70]: and the best performance is assumed to be at the max_resource?

File: flaml/tune/tune.py [hunk 13]
```diff
@@ -205,21 +207,28 @@ def compute_with_config(config):
...
-            prune_attr; only valid if prune_attr is not in space.
+        resource_attr: A string to specify the resource dimension used by
+            the scheduler via "scheduler".
+        min_resource: A float of the minimal resource to use for the resource_attr.
+        max_resource: A float of the maximal resource to use for the resource_attr.
         reduction_factor: A float of the reduction factor used for incremental
             pruning.
-        report_intermediate_result: A boolean of whether intermediate results
-            are reported. If so, early stopping and pruning can be used.
+        scheduler: A scheduler for executing the experiment. Can be 'flaml', 'asha'
```
Comment (sonichi) [inline flaml/tune/tune.py:216]: Explain the default behavior.

File: flaml/tune/tune.py [hunk 14]
```diff
@@ -295,6 +304,19 @@ def compute_with_config(config):
     from ..searcher.blendsearch import BlendSearch
 
     if search_alg is None:
+        flaml_scheduler_resource_attr = (
+            flaml_scheduler_min_resource
+        ) = flaml_scheduler_max_resource = flaml_scheduler_reduction_factor = None
+        if scheduler == "flaml":
```
Comment (sonichi) [inline flaml/tune/tune.py:310]:
```suggestion
        if scheduler in (None, "flaml"):

```

File: flaml/searcher/blendsearch.py [hunk 15]
```diff
@@ -91,17 +91,10 @@ def __init__(
...
-                When prune_attr is in space, it is a hyperparameter, e.g.,
-                    'n_iters', and the best value is unknown.
-                When prune_attr is not in space, it is a resource dimension,
-                    e.g., 'sample_size', and the peak performance is assumed
-                    to be at the max_resource.
-            min_resource: A float of the minimal resource to use for the
-                prune_attr; only valid if prune_attr is not in space.
-            max_resource: A float of the maximal resource to use for the
-                prune_attr; only valid if prune_attr is not in space.
+            resource_attr: A string to specify the resource dimension and and the best
```
Comment (sonichi) [inline flaml/searcher/blendsearch.py:94]:
```suggestion
            resource_attr: A string to specify the resource dimension and the best
```

File: flaml/tune/tune.py [hunk 16]
```diff
@@ -12,11 +12,13 @@
 
     assert ray_version >= "1.0.0"
     from ray.tune.analysis import ExperimentAnalysis as EA
+    from ray.tune.schedulers import TrialScheduler
```
Comment (sonichi) [inline flaml/tune/tune.py:15]: Where is TrialScheduler used?

Comment (qingyun-wu) [reply flaml/tune/tune.py]: It was used in the type of the `scheduler` argument. But I removed the type. Removed the import as well now.

**Review by sonichi — APPROVED** [review]

**Pull request merged**